sound/soc/sof/ipc4-pcm.c

```diff
@@ -37,72 +64,77 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
 	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
-	struct snd_sof_widget *pipeline_widget;
-	struct snd_soc_dapm_widget_list *list;
-	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_pipeline_trigger_info *pipeline_info;
+	struct ipc4_pipeline_set_state_data data;
+	struct snd_sof_widget *pipe_widget;
 	struct sof_ipc4_pipeline *pipeline;
-	struct snd_sof_widget *swidget;
 	struct snd_sof_pcm *spcm;
-	int ret = 0;
-	int num_widgets;
+	int ret, i, j;
 
 	spcm = snd_sof_find_spcm_dai(component, rtd);
 	if (!spcm)
 		return -EINVAL;
 
-	list = spcm->stream[substream->stream].list;
+	pipeline_info = &spcm->stream[substream->stream].pipeline_info;
+	data.count = 0;
 
-	for_each_dapm_widgets(list, num_widgets, widget) {
-		swidget = widget->dobj.private;
-
-		if (!swidget)
-			continue;
+	/*
+	 * IPC4 requires pipelines to be triggered in order starting at the sink and
+	 * walking all the way to the source. So add the pipeline instance ID's starting with the
+	 * last one in the trigger list. Skip those pipelines that are already in the requested
+	 * state
+	 */
+	for (i = pipeline_info->count - 1; i >= 0; i--) {
+		pipe_widget = pipeline_info->pipelines[i];
+		pipeline = pipe_widget->private;
 
 		/*
-		 * set pipeline state for both FE and BE pipelines for RUNNING state.
-		 * For PAUSE/RESET, set the pipeline state only for the FE pipeline.
+		 * skip triggering pipelines that are already in the requested state. This will
+		 * be true for backend pipelines that are triggered in the BE DAI trigger.
 		 */
-		switch (state) {
-		case SOF_IPC4_PIPE_PAUSED:
-		case SOF_IPC4_PIPE_RESET:
-			if (!WIDGET_IS_AIF(swidget->id))
-				continue;
-			break;
-		default:
-			break;
-		}
+		if (pipeline->state != state)
```

Comment:
As next state it is setting is ```SOF_IPC4_PIPE_PAUSED```, we should not skip ```pipeline->state != state``` pipelines.

Reply:
@libinyang you're talking a hypothetical scenario. How is it possible that you have pipeline 2 running while both pipeline 1 and 3 are paused. Would that pipeline immediately run into an xrun?

Reply:
@ranj063 OK. If there is no such scenario, it's OK to me.